Organisms/TodoCardの作成2

caramelribon opened the pull request:
## Organisms/TodoCardの作成

Atoms/Inputの方で修正を加えたので、新しくブランチを切りました
前回までのTodoCardのPRは"Organisms/TodoCardの作成1"の方を見て頂ければと思います
・Enter押したときに、エラーができていたので、atoms/inputの方で、修正したのを適用させています
・タスクを追加したときに、追加したタスクにだけdefaultIsEditingが適用されるように修正しました

- [x] お手本通りの見た目であるか　ホバー時の挙動、レスポンシブ対応など要確認
- [x] ブランチ名、ファイル名、クラス名、インデントなど、コーディング規則を逸脱していないか
- [x] 記述の重複はないか

caramelribon committed:
Organisms/TodoCardの作成

Enter押したときに、エラーができていたので、atoms/inputの方で、修正したのを適用させています

タスクを追加したときに、追加したタスクにだけdefaultIsEditingが適用されるように修正しました

Higashiji commented on src/components/Organisms/TodoCard/index.jsx:
```diff
@@ -0,0 +1,98 @@
+  // task.stateがTODOのものだけ抽出する
+  todoStateTask = todoTask.filter((task) => {
+    return task.state === "TODO";
+  });
```
```suggestion
const  todoStateTask = todoTask.filter((task) => {
    return task.state === "TODO";
  });
```
11行目の変数宣言と分ける意味がないのでここにまとめて書きましょう。

プログラミング中に手続型と関数型という考え方があり、reactは関数型の考え方に基づいたライブラリだと言われています。
関数型の考えに則れば、変数への再代入は避けたほうがいいものとなります。

以下は少し高度な記事ですが、参考になりそうです。
https://zenn.dev/tockri/books/dcaf6c55e64448/viewer/react

Higashiji commented on src/components/Organisms/TodoCard/index.jsx:
```diff
@@ -0,0 +1,98 @@
+  };
+
+  return (
+    <StyledTodoCard>
+      <AddTaskButton onClick={addTask} />
+      <StyledTodoList>{todoTaskList}</StyledTodoList>
+    </StyledTodoCard>
```
```suggestion
    <StyledTodoCard>
      <AddTaskButton onClick={addTask} />
      <StyledTodoList>
      　{ todoStateTask.map((task, index) =>  (
      <StyledTodoListItem key={index}>
        <Task
          key={task.id}
          defalutValue={task.name}
          defalutIsEditing={
            taskAddOREdit ? task.id === todoTask.length - 1 : false
          }
          onEditComplete={(taskName) => handleEditComplete(task.id, taskName)}
          onTaskComplete={() => handleTaskComplete(task.id)}
        />
      </StyledTodoListItem>
      )}
      </StyledTodoList>
    </StyledTodoCard>
```
コンポーネント内でJSXの記述が2つに散っているのが気になるので、このように一つにまとめてもいいと思います。
（ここは感性の問題な気もするので修正しなくても大丈夫です。）

Higashiji commented on src/components/Organisms/TodoCard/index.jsx:
```diff
@@ -0,0 +1,98 @@
+  let todoStateTask = [];
+
+  const handleTaskComplete = (id) => {
+    let completedTask = todoTask.map((task) => {
```
completedTaskには再代入がないのでconstでいいと思います。

Higashiji commented on src/components/Organisms/TodoCard/index.jsx:
```diff
@@ -0,0 +1,98 @@
+  };
+
+  const handleEditComplete = (id, taskName) => {
+    let edittedTask = [];
```
editedTaskが正しいかと！

Higashiji commented on src/components/Organisms/TodoCard/index.jsx:
```diff
@@ -0,0 +1,98 @@
+import COLOR from "../../../variables/color";
+
+const TodoCard = () => {
+  const [todoTask, setTodoTask] = useState([]);
```
todoTaskには state: "TODO" のタスクのみが入ると予想する人がいると思います（自分がそうでした）
単に `taskList` や `tasks` というstate名にすると、そのような誤解が防げると思います。

Higashiji commented on src/components/Organisms/TodoCard/index.jsx:
```diff
@@ -0,0 +1,98 @@
+
+const TodoCard = () => {
+  const [todoTask, setTodoTask] = useState([]);
+  const [taskAddOREdit, setAddOREdit] = useState(false);
```
```suggestion
  const [taskAddOrEdit, setAddOrEdit] = useState(false);
```
厳密にはキャメルケースだとこうなると思います。

Higashiji commented on src/components/Organisms/TodoCard/index.jsx:
```diff
@@ -0,0 +1,98 @@
+        <Task
+          key={task.id}
+          defalutValue={task.name}
+          defalutIsEditing={
+            taskAddOREdit ? task.id === todoTask.length - 1 : false
+          }
```
```suggestion
          defalutIsEditing={true}
```
手元で試した限りこれで問題なさそうです。
（初回レンダリング（マウント時）にのみ関係があるpropsなので、親コンポーネントの方で状態管理する必要がない気がします。）

Higashiji replied:
それと、　defalut -> default だと思います。

Higashiji commented on src/components/Organisms/TodoCard/index.jsx:
```diff
@@ -0,0 +1,98 @@
+    const idNum = todoTask.length;
+    setTodoTask((todoTask) => [
+      ...todoTask,
+      { id: idNum, name: "", state: "TODO" },
```
idは結局　todoTaskにおけるindexと同じなので、idに代入せずindexとして使うほうがかえって可読性が上がる気もします。（動作ベースで判断するので、修正しなくても大丈夫です。）

Higashiji reviewed (commented):
typo 対策に code spell checker という拡張がおすすめです！
https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker

caramelribon committed:
Organisms/TodoCard

変数名の修正

再代入がないものはconstにする

誤解を招かないようにstate名の変更

idを削除してindexを用いて配列の操作をするように記述しました

caramelribon committed:
Organisms/TodoCard

・タスクが完了したときの動作の修正

タスクが完了した場合、そのタスクのデータを配列から削除するように修正しました

・タスク追加とタスク編集時のstateの削除

タスクを追加したときに、defaultEditingを渡すために、状態管理していたuseStateが必要なくなったので削除しました

caramelribon commented:
## Organisms/TodoCardの作成

・タスクが完了したときの動作を修正しました。
・defaultEditingを渡すために使用していたtaskAddOrEditのuseStateが必要なくなったので削除しました。

- [x] お手本通りの見た目であるか　ホバー時の挙動、レスポンシブ対応など要確認
- [x] ブランチ名、ファイル名、クラス名、インデントなど、コーディング規則を逸脱していないか
- [x] 記述の重複はないか

Higashiji commented on src/components/Organisms/TodoCard/index.jsx:
```diff
@@ -0,0 +1,82 @@
+  const [taskList, setTaskList] = useState([]);
+
+  // タスクのStateが"TODO"のものを抽出
+  const todoStateTask = taskList.filter((task) => {
+    return task.state === "TODO";
+  });
+
+  // タスクを追加したときの関数(AddTaskButtonを押したとき)
+  const addTask = () => {
+    setTaskList((taskList) => [...taskList, { name: "", state: "TODO" }]);
+  };
```
もはやstate === "TODO"のタスクしか存在しないのでtodoStateTaskを生成する必要はない気もします！
動作に影響はないのであっても大丈夫です。

Higashiji reviewed (approved):
問題ないと思います！
